### PR TITLE
Geonode client

### PR DESCRIPTION
**Pull request by myarjunar**
## Description
Add general functionality to the manager dialog. ( #2 #9 #11 ) 
Add geonode browser panel ( still cannot populate layers/maps to the browser ). ( #3  )
Add layer from geonode manager dialog to the map.

Browser panel:
geonode-browser-panel

Create new geonode connection:
create-new-geonode-connection

Save geonode connection:
save-geonode-connection

Load geonode connection from exported connection:
load-geonode-connection

Delete geonode connection:
delete-geonode-connection

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by timlinux** [comment]
That is looking really great @myarjunar  and @ismailsunni !

**Comment by Gustry** [comment]
You should update master at the same time. The PR is hard to read now.
Maybe github will still display old commits from QGIS. A rebase may be needed.

File: src/app/geonode/qgsgeonodesourceselect.cpp [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+  }
+
+  mAddButton = new QPushButton( tr( "&Add" ) );
+  mAddButton->setEnabled( false );
+
+  buttonBox->addButton( mAddButton, QDialogButtonBox::ActionRole );
+
+  populateConnectionList();
+
+  connect( buttonBox, SIGNAL( rejected() ), this, SLOT( reject() ) );
```
Comment (Gustry) [inline src/app/geonode/qgsgeonodesourceselect.cpp]: This is old style connect. We need to move to the new style.

File: src/app/geonode/qgsgeonodesourceselect.cpp [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+  connect( btnSave, SIGNAL( clicked() ), this, SLOT( saveGeonodeConnection() ) );
+  connect( btnLoad, SIGNAL( clicked() ), this, SLOT( loadGeonodeConnection() ) );
+  connect( lineFilter, SIGNAL( textChanged( QString ) ), this, SLOT( filterChanged( QString ) ) );
+  connect( treeView, SIGNAL( clicked( QModelIndex ) ), this, SLOT( treeViewSelectionChanged( QModelIndex ) ) );
+
+  mItemDelegate = new QgsGeonodeItemDelegate( treeView );
+  treeView->setItemDelegate( mItemDelegate );
+
+  mModel = new QStandardItemModel();
+  mModel->setHorizontalHeaderItem( MODEL_IDX_TITLE, new QStandardItem( QStringLiteral( "Title" ) ) );
```
Comment (Gustry) [inline src/app/geonode/qgsgeonodesourceselect.cpp]: Missing translation ?

**Comment by ismailsunni** [comment]
Oops, the commit from Tim and I for fixing the conflict got merged in one of the commit.

**Comment by Gustry** [comment]
Is-it not fine ?

**Comment by ismailsunni** [comment]
@Gustry it's fine. Nothing is lose, beside Tim's name in the list of this PR commiters.

File: src/app/geonode/qgsgeonodenewconnection.cpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    if ( !authcfg.isEmpty() )
+    {
+      tabAuth->setCurrentIndex( tabAuth->indexOf( mAuthConfigSelect ) );
+    }
+  }
+
+  // Adjust height
+  int w = width();
+  adjustSize();
+  resize( w, height() );
```
Comment (Gustry) [inline src/app/geonode/qgsgeonodenewconnection.cpp]: Can't we use width() here ? Or do you need `w` ?

File: src/app/geonode/qgsgeonodenewconnection.cpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+  buttonBox->button( QDialogButtonBox::Ok )->setDisabled( true );
+  connect( txtName, SIGNAL( textChanged( QString ) ), this, SLOT( okButtonBehavior( QString ) ) );
+  connect( txtUrl, SIGNAL( textChanged( QString ) ), this, SLOT( okButtonBehavior( QString ) ) );
+  connect( btnConnect, SIGNAL( clicked() ), this, SLOT( testConnection() ) );
```
Comment (Gustry) [inline src/app/geonode/qgsgeonodenewconnection.cpp]: You need to remove these old style connect

File: src/providers/ows/qgsowsdataitems.cpp [hunk 5]
```diff
@@ -271,4 +280,80 @@ QGISEXTERN QDialog *selectWidget( QWidget *parent, Qt::WindowFlags fl )
...
+
+void QgsGeoNodeRootItem::newConnection()
+{
+  QgsGeoNodeNewConnection *nc = new QgsGeoNodeNewConnection( nullptr );
+
+  if ( nc->exec() )
+  {
+    refresh();
+  }
+}*/
```
Comment (Gustry) [inline src/providers/ows/qgsowsdataitems.cpp]: Still useful ?

Comment (myarjunar) [inline src/providers/ows/qgsowsdataitems.cpp]: yes, I leave it there as a reminder :sweat_smile:

File: src/app/geonode/qgsgeonodenewconnection.cpp (hunk 3 above)
Comment (myarjunar) [inline src/app/geonode/qgsgeonodenewconnection.cpp]: somehow if we use `width()` directly, the size is not as what I'm expecting.

File: src/app/qgisapp.cpp [hunk 6]
```diff
@@ -4464,6 +4464,8 @@ void QgisApp::addGeonodeLayer()
   }
   connect( geonodes, SIGNAL( addRasterLayer( QString const &, QString const &, QString const & ) ),
            this, SLOT( addRasterLayer( QString const &, QString const &, QString const & ) ) );
+  connect( geonodes, SIGNAL( addWfsLayer( QString, QString, QString ) ),
```
Comment (Gustry) [inline src/app/qgisapp.cpp]: This an old style connect. You need to use the new-style, QGIS doesn't allow new "old" style anymore.

Comment (ismailsunni) [inline src/app/qgisapp.cpp]: Ah shame, I didn't check that one. I was focusing to fix the new slot. Ok, I will fix it. Thanks @Gustry

**Comment by Gustry** [comment]
Can we close this PR? as it's not on this branch anyway and there is the PR on the QGIS repo now

**Pull request closed without merging**